tzrec/acc/aot_utils.py

```diff
@@ -0,0 +1,139 @@
+    if half_to_float:
+        assert x.dtype == torch.half
+    computation_dtype, result_dtype = prims_utils.elementwise_dtypes(
+        x, type_promotion_kind=prims_utils.ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
+    )
+    x = x.to(computation_dtype)
+    x_max = torch.max(x, dim, keepdim=True).values
+    unnormalized = torch.exp(x - x_max)
+    result = unnormalized / torch.sum(unnormalized, dim, keepdim=True)
+    if not half_to_float:
+        result = result.to(result_dtype)
+    return result
+
+
+def export_model_aot(
+    model: nn.Module, data: Dict[str, torch.Tensor], save_dir: str
+) -> torch.export.ExportedProgram:
+    """Export aot model.
+
+    Args:
+        model (nn.Module): the model
+        data (Dict[str, torch.Tensor]): the test data
+        save_dir (str): model save dir
+    """
+    gm = symbolic_trace(model)
+    with open(os.path.join(save_dir, "gm.code"), "w") as f:
+        f.write(gm.code)
+    # with open(os.path.join(save_dir, "gm.graph"), "w") as f:
+    #     f.write(gm.graph.print_tabular())
+
+    gm = gm.cuda()
+
+    print(gm)
+
+    batch = Dim("batch")
+    dynamic_shapes = {}
+    for key in data:
+        if key.endswith(".lengths"):
+            if data[key].shape[0] == 1:
+                logger.info("uniq user sparse fea %s length=1" % key)
+                dynamic_shapes[key] = {}
+            else:
+                dynamic_shapes[key] = {0: batch}
+        elif key == "batch_size":
+            dynamic_shapes[key] = {}
+        elif data[key].dtype == torch.float32 and "__" not in key:
+            if data[key].shape[0] == 1:
+                logger.info("uniq user dense_fea=%s shape=%s" % (key, data[key].shape))
+                dynamic_shapes[key] = {}
+            else:
+                logger.info("dense_fea=%s shape=%s" % (key, data[key].shape))
+                dynamic_shapes[key] = {0: batch}
+        elif (
+            data[key].dtype == torch.float32 and "__" in key and data[key].shape[0] == 1
+        ):
+            logger.info("uniq seq_dense_fea=%s shape=%s" % (key, data[key].shape))
+            dynamic_shapes[key] = {}
+        else:
+            tmp_val_dim = Dim(key.replace(".", "__") + "__batch", min=0)
+            # to handle torch.export 0/1 specialization problem
+            if data[key].shape[0] < 2:
+                data[key] = F.pad(
+                    data[key],
+                    [0, 2] + [0, 0] * (len(data[key].shape) - 1),
+                    mode="constant",
+                )
+            dynamic_shapes[key] = {0: tmp_val_dim}
+
+    exported_gm = torch.export.export(
+        gm, args=(data,), dynamic_shapes=(dynamic_shapes,)
+    )
+    print(exported_gm)
+
+    export_path = os.path.join(save_dir, "exported_gm.code")
+    with open(export_path, "w") as fout:
+        fout.write(str(exported_gm))
+
+    exported_gm_path = os.path.join(save_dir, "debug_exported_gm.py")
+    with open(exported_gm_path, "w") as fout:
+        fout.write(str(exported_gm))
```

Comment:
the exported_gm.code == debug_exported_gm,py ,they all save the str(exported_gm)